README.md
# lvmloader

![Versions](https://img.shields.io/badge/python->3.7-blue)
[![Documentation Status](https://readthedocs.org/projects/sdss-lvmloader/badge/?version=latest)](https://sdss-lvmloader.readthedocs.io/en/latest/?badge=latest)
[![Travis (.org)](https://img.shields.io/travis/sdss/lvmloader)](https://travis-ci.org/sdss/lvmloader)
[![codecov](https://codecov.io/gh/sdss/lvmloader/branch/main/graph/badge.svg)](https://codecov.io/gh/sdss/lvmloader)

Package for ingesting LVM data into the lvmdb at Utah


CHANGELOG.rst
.. _lvmloader-changelog:

==========
Change Log
==========

* First change goes here.



tests/test_main.py
# encoding: utf-8
#
# main.py

from pytest import mark

from lvmloader.main import math


class TestMath(object):
    """Tests for the ``math`` function in main.py."""

    @mark.parametrize(('arg1', 'arg2', 'operator', 'result'),
                      [(1, 2, '+', 3), (2, 2, '-', 0), (3, 5, '*', 15), (10, 2, '/', 5)])
    def test_math(self, arg1, arg2, operator, result):

        assert math(arg1, arg2, arith_operator=operator) == result


python/lvmloader/exceptions.py
# !usr/bin/env python
# -*- coding: utf-8 -*-
#
# Licensed under a 3-clause BSD license.
#
# @Author: Brian Cherinka
# @Date:   2017-12-05 12:01:21
# @Last modified by:   Brian Cherinka
# @Last Modified time: 2017-12-05 12:19:32

from __future__ import absolute_import, division, print_function


class LvmloaderError(Exception):
    """A custom core Lvmloader exception"""

    def __init__(self, message=None):

        message = 'There has been an error' \
            if not message else message

        super(LvmloaderError, self).__init__(message)


class LvmloaderNotImplemented(LvmloaderError):
    """A custom exception for not yet implemented features."""

    def __init__(self, message=None):

        message = 'This feature is not implemented yet.' \
            if not message else message

        super(LvmloaderNotImplemented, self).__init__(message)


class LvmloaderAPIError(LvmloaderError):
    """A custom exception for API errors"""

    def __init__(self, message=None):
        if not message:
            message = 'Error with Http Response from Lvmloader API'
        else:
            message = 'Http response error from Lvmloader API. {0}'.format(message)

        super(LvmloaderAPIError, self).__init__(message)


class LvmloaderApiAuthError(LvmloaderAPIError):
    """A custom exception for API authentication errors"""
    pass


class LvmloaderMissingDependency(LvmloaderError):
    """A custom exception for missing dependencies."""
    pass


class LvmloaderWarning(Warning):
    """Base warning for Lvmloader."""


class LvmloaderUserWarning(UserWarning, LvmloaderWarning):
    """The primary warning class."""
    pass


class LvmloaderSkippedTestWarning(LvmloaderUserWarning):
    """A warning for when a test is skipped."""
    pass


class LvmloaderDeprecationWarning(LvmloaderUserWarning):
    """A warning for deprecated features."""
    pass


python/lvmloader/ingest/drp.py
# !/usr/bin/env python
# -*- coding: utf-8 -*-
#

import itertools
import pathlib
import time

import astropy.units as u
import numpy as np
from astropy.coordinates import SkyCoord
from astropy.io import fits
from astropy.table import Table
from sdssdb.sqlalchemy.lvmdb import database, drp
from sqlalchemy import insert, select

from lvmloader.bundle import LVMBundle, MangaBundle


class Timer:

    def __init__(self, message: str = '', verbose: bool = None):
        self.message = message
        self.verbose = verbose
        self.start_time = None
        self.end_time = None
        self.elapsed = None

    def __enter__(self):
        self.start_time = time.perf_counter()

    def __exit__(self, *exc_info):
        self.end_time = time.perf_counter()
        self.elapsed = self.end_time - self.start_time
        if self.verbose:
            print(f'{self.message} - Elapsed time [sec]: {self.elapsed}')


class LVMLoader:

    def __init__(self, filename: str, release: str = None, verbose: bool = None,
                 manga: bool = False):
        self.filepath = pathlib.Path(filename).resolve()
        self.filename = self.filepath.name
        self.drpver = None
        self.dapver = None
        self.release = release
        self.session = database.Session()
        self.verbose = verbose
        self.manga = manga

    def __repr__(self):
        return f'<LVMLoader(file="{self.filename}")'

    def _close(self):
        self.session.close()

    def __del__(self):
        self._close()

    def __enter__(self):
        return self

    def __exit__(self, *exc_info):
        self._close()

    def load_into_db(self):

        print(f'Starting load: {self.filename}')
        with Timer(f'Loading file: {self.filename}', verbose=self.verbose):
            with fits.open(self.filepath) as hdu:
                self.set_params(hdu)

                with database.Session() as session, session.begin():
                    self.session = session
                    with Timer('Loading Metadata', verbose=self.verbose):
                        self.load_rss()
                        self.load_wavelength(hdu)
                        self.load_header(hdu)
                        self.load_obsinfo(hdu)
                with Timer('Loading pixels', verbose=self.verbose):
                    self.load_fibers(hdu)
        self._close()

    def set_params(self, hdu):
        # temporarily using MaNGA
        self.drpver = hdu[0].header.get('VERSDRP2')
        self.dapver = hdu[0].header.get('VERSDAP')
        self.tileid = int(hdu[0].header['PLATEIFU'].replace('-', ''))
        self.ra = hdu[0].header.get('IFURA')
        self.dec = hdu[0].header.get('IFUDEC')
        self.ring = 7 if str(hdu[0].header['IFUDSGN']).startswith('127') else 6

    def get_footprint(self):
        if self.manga:
            b = MangaBundle(ra=self.ra, dec=self.dec, ring=self.ring)
        else:
            b = LVMBundle(ra=self.ra, dec=self.dec, ring=self.ring)
        return b.hexagon.to_table().as_array().tolist()

    def load_rss(self):
        q = select(drp.RSS).where(drp.RSS.tileid == self.tileid)
        rss = self.session.scalars(q).one_or_none()

        if not rss:
            foot = self.get_footprint()
            rss = drp.RSS(tileid=self.tileid, ifura=self.ra, ifudec=self.dec, footprint=foot)
            rss.pipeline = self.load_pipeline()
            self.session.add(rss)

        return rss

    def load_wavelength(self, hdu):

        if ww := self.session.scalars(select(drp.Wavelength)).all():
            return ww

        wave = hdu['WAVE'].data.tolist()
        nwave = len(wave)
        ww = [dict(zip(['index', 'value'], row)) for row in zip(np.arange(nwave).tolist(), wave)]
        self.session.execute(insert(drp.Wavelength), ww)

    def get_header(self, hdu):
        hdr = hdu['PRIMARY'].header
        rss = self.load_rss()
        return [dict(zip(['index', 'key', 'value', 'comment', 'rss_pk'], row))
                for row in zip(np.arange(len(hdr)).tolist(), hdr.keys(), hdr.values(),
                               hdr.comments, itertools.repeat(rss.pk))]

    def load_header(self, hdu):

        rss = self.load_rss()
        if self.session.scalars(select(drp.Header).where(drp.Header.rss == rss)).first():
            return
        data = self.get_header(hdu)
        self.session.execute(insert(drp.Header), data)

    def load_pipeline(self, label: str = 'DRP'):

        version = self.drpver if label == 'DRP' else self.dapver if label == 'DAP' else None

        q = select(drp.Pipeline).where(drp.Pipeline.version == version)
        pipe = self.session.scalars(q).one_or_none()

        if not pipe:
            pipe = drp.Pipeline(version=self.drpver, label=label, release=self.release)
            self.session.add(pipe)

        return pipe

    def get_obsinfo(self, hdu):
        rss = self.load_rss()

        cols = ['EXPNUM', 'MJD', 'EXPTIME', 'MGDPOS', 'MGDRA', 'MGDDEC',
                'SEEING', 'BLUESN2', 'REDSN2', 'DRP2QUAL']

        tt = Table(hdu['OBSINFO'].data)
        df = tt[cols].to_pandas()
        df.columns = df.columns.str.lower()

        dmap = {'C': 0, 'N': 1, 'E': 2, 'S': 3}
        df['mgdpos'] = df['mgdpos'].apply(lambda x: dmap[x.strip()])
        df['expnum'] = df['expnum'].astype(int)
        df.insert(0, 'rss_pk', rss.pk)

        # cols = ['rss_pk', 'expnum', 'mjd', 'exptime', 'nexp', 'dpos',
        #         'dra', 'ddec', 'airmass', 'seeing', 'bluesn2', 'redsn2'
        #         'nirsn2', 'drp2qual']

        df = df.rename(columns={'mgdpos': 'dpos', 'mgdra': 'dra', 'mgddec': 'ddec'})

        return df.to_dict(orient='records')

    def load_obsinfo(self, hdu):
        rss = self.load_rss()
        if self.session.scalars(select(drp.ObsInfo).where(drp.ObsInfo.rss == rss)).first():
            return
        data = self.get_obsinfo(hdu)
        self.session.execute(insert(drp.ObsInfo), data)

    def convert_xy_radec(self, x, y):
        s = SkyCoord(self.ra, self.dec, unit=u.degree)
        radecs = s.spherical_offsets_by(x * u.arcsec, y * u.arcsec)
        return zip(*radecs.to_table().as_array().tolist())

    def get_fiber_data(self, hdu):
        xp = hdu['XPOS'].data
        yp = hdu['YPOS'].data
        nrows, nwave = xp.shape
        nexp = len(hdu['OBSINFO'].data)
        nfiber = int(nrows / nexp)
        wrange = np.arange(nwave).tolist()
        widx = int(np.median(wrange))

        rss = self.load_rss()
        rsspk = itertools.repeat(rss.pk)

        # TODO look up real fibers pk
        fibpk = np.tile(np.arange(nfiber) + 1, nexp).tolist()

        # look up obsinfo
        q = select(drp.ObsInfo.pk).where(drp.ObsInfo.rss == rss)
        if obsinfo := self.session.scalars(q).all():
            obspk = np.repeat(obsinfo, nfiber).tolist()
        else:
            obspk = np.repeat(None, nrows).tolist()

        yidx = np.arange(nrows).tolist()
        xpos = xp[:, wrange].ravel().tolist()
        ypos = yp[:, wrange].ravel().tolist()

        # get wave_pk
        wp = self.session.scalars(select(drp.Wavelength).where(drp.Wavelength.index == widx)).one()
        wave_pk = np.tile(wp.pk, nrows).tolist()

        # get ra/dec
        ra, dec = self.convert_xy_radec(xpos, ypos)

        cols = ['rss_pk', 'fibers_pk', 'obsinfo_pk',
                'yidx', 'xpos', 'ypos', 'wave_pk', 'ra',
                'dec']

        return [dict(zip(cols, row))
                for row in zip(rsspk, fibpk, obspk, yidx, xpos,
                               ypos, wave_pk, ra, dec)]

    def load_fibers(self, hdu):

        rss = self.load_rss()
        if self.session.scalars(select(drp.RSSFiber).where(drp.RSSFiber.rss == rss)).first():
            return

        data = self.get_fiber_data(hdu)

        with database.engine.begin() as conn:
            ins = drp.RSSFiber.__table__.insert()
            conn.execute(ins, data)


# files = list(pathlib.Path('/Users/Brian/Work/Manga/redux/v3_1_1/').rglob("*[9,11]*/*/*RSS*"))
def load_lvm_files(files, manga=None):
    for file in files:
        ll = LVMLoader(file, verbose=True, manga=manga)
        ll.load_into_db()


python/lvmloader/bundle/bundle.py
# !usr/bin/env python
# -*- coding: utf-8 -*-
#
import os

import astropy.units as u
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from astropy.coordinates import SkyCoord
from astropy.io import fits
from astropy.visualization.wcsaxes import SphericalCircle
from astropy.wcs import WCS
from regions import PolygonSkyRegion


__all__ = ['MangaBundle', 'LVMBundle']


class BaseBundle(object):
    """The location, size, and shape of a LVM IFU bundle.

    A bundle of a given size at the RA/Dec coordinates.

    Setting envelope creates the hexagon at the outer boundry of the fibers,
    otherwise the hexagon runs through the centers of the outer fibers.

    Parameters:
        ra (float):
            The Right Ascension of the target
        dec (float):
            The Declination of the target
        ring (int):
            The ring size of the IFU, e.g. 25 = 1801 fibers
        use_envelope (bool):
            Expands the hexagon area to include an envelope surrounding the hex border
        local (bool):
            If True, grabs the fiducial metrology file from a local LVMCORE

    Attributes:
        fibers (SkyCoord):
            A SkyCoord array of IFU fiber coordinates
        skies (SkyCoord):
            An Nx2 array of sky fiber coordinates.  Default is unloaded.
            Use :meth:`get_sky_coordinates` to load.
        hexagon (SkyCoord):
            A SkyCoord array of coordinates defining the hexagon vertices
        sky_region (PolygonSkyRegion):
            An Astropy SkyRegion
    """

    def __init__(self, ra: float, dec: float, ring: int = 25, use_envelope: bool = True,
                 local: bool = None, **kwargs):
        self.ra = ra
        self.dec = dec
        self.coord = SkyCoord(self.ra, self.dec, unit=u.degree)
        self.fiber_diameter = self.pixel_scale * u.arcsec

        self.simbmap = None

        self.size = self.total_fiber_num(ring)
        self.rings = np.arange(1, ring + 1)
        self.totals = self.total_fiber_num(self.rings)

        self.hexagon = None
        self.sky_region = None

        if self.size not in self.totals:
            return

        self.simbmap = self._get_simbmap_file(local=local)

        # get the fiber coordinates
        self.fibers = self.get_fiber_coordinates(size=self.size)

        # get the hexagon coordinates and create a sky region
        self.hexagon = self._calculate_hexagon(use_envelope=use_envelope)
        self.sky_region = PolygonSkyRegion(vertices=self.hexagon)

    def __repr__(self):
        return '<Bundle (ra={0}, dec={1}, ifu={2})>'.format(self.ra, self.dec, self.size)

    @staticmethod
    def total_fiber_num(ring: int) -> int:
        """ compute the total fiber number within a ring id """
        return 3 * ((ring - 1) ** 2) + 3 * (ring - 1) + 1

    def _get_simbmap_file(self, local: bool = None) -> pd.DataFrame:
        ''' Retrieves the metrology file locally or remotely

        Reads in fresh metrology data or grabs it from
        cache if available

        Parameters:
            local (bool):
                If True, does a local system check

        Returns:
            The metrology object data
        '''

        # create the file path
        rel_path = f'metrology/{self.simfile}'
        if local:
            if 'LVMCORE_DIR' not in os.environ:
                raise ValueError("LVMCORE_DIR environment variable must be set.")
            base_path = os.environ['LVMCORE_DIR']
        else:
            base_path = u'https://raw.githubusercontent.com/sdss/lvmcore/master'

        self.simbmap_file = os.path.join(base_path, rel_path)
        return pd.read_table(self.simbmap_file, sep=',', header=21)

    def get_fiber_coordinates(self, size: int = None):
        """ Returns the RA, Dec coordinates for each fiber.

        Parameters:
            size (int):
                The IFU size.  This extracts only the fibers for the given size

        Returns:
            An Nx3 numpy array, each row containing [fiberid, RA, Dec]
        """

        fibers = self.coord.spherical_offsets_by(
            self.simbmap['raoff'] * u.arcsec, self.simbmap['decoff'] * u.arcsec)

        # extract only the fibers for the specified IFU size
        size = size or self.size
        fibers = fibers[:size]

        return fibers

    def _calculate_hexagon(self, use_envelope: bool = True):
        """ Calculates the vertices of the bundle hexagon. """

        simbmap = self.simbmap[:self.size].copy()

        # assign the row axes based on the original angle
        row, col = ('raoff', 'decoff') if self.angle != 0 else ('decoff', 'raoff')

        # identify middle, top, bottom row slices
        middleRow = simbmap[simbmap[row].round(2) == 0]
        topRow = simbmap[simbmap[row].round(2).max() - simbmap[row].round(2) == 0]
        bottopRow = simbmap[simbmap[row].round(2).min() - simbmap[row].round(2) == 0]

        if len(topRow) < max(self.rings) or len(bottopRow) < max(self.rings):
            print('Warning, did not select all fibers in top/bottom row')

        # identify hexagon vertices
        vertice0 = middleRow[middleRow[col] == np.max(middleRow[col])]
        vertice3 = middleRow[middleRow[col] == np.min(middleRow[col])]

        vertice1 = topRow[topRow[col] == np.max(topRow[col])]
        vertice2 = topRow[topRow[col] == np.min(topRow[col])]

        vertice5 = bottopRow[bottopRow[col] == np.max(bottopRow[col])]
        vertice4 = bottopRow[bottopRow[col] == np.min(bottopRow[col])]

        hexagonOff = np.array(
            [[vertice0['raoff'].values[0], vertice0['decoff'].values[0]],
             [vertice1['raoff'].values[0], vertice1['decoff'].values[0]],
             [vertice2['raoff'].values[0], vertice2['decoff'].values[0]],
             [vertice3['raoff'].values[0], vertice3['decoff'].values[0]],
             [vertice4['raoff'].values[0], vertice4['decoff'].values[0]],
             [vertice5['raoff'].values[0], vertice5['decoff'].values[0]]])

        # This array increases the area of the hexagon so that it is an
        # envelope of the bundle.
        if use_envelope:
            halfFibre = self.fiber_diameter.value / 2.
            hexagonExtra = np.array(
                [[halfFibre, 0.0],
                 [halfFibre, halfFibre],
                 [-halfFibre, halfFibre],
                 [-halfFibre, 0.0],
                 [-halfFibre, -halfFibre],
                 [halfFibre, -halfFibre]])

            if self.angle != 0:
                angle_rad = self.angle * np.pi / 180
                rot = np.array([[np.cos(angle_rad), -np.sin(angle_rad)],
                                [-np.sin(angle_rad), np.cos(angle_rad)]])
                hexagonExtra = np.dot(hexagonExtra, rot)

            hexagonOff += hexagonExtra

        self._hexagon_offsets = hexagonOff
        r, d = zip(*hexagonOff)
        return self.coord.spherical_offsets_by(r * u.arcsec, d * u.arcsec)

    def _rotate_data(self, data, angle):
        angle_rad = angle * np.pi / 180

        x = data[:, 0]
        y = data[:, 1]

        newx = x * np.cos(angle_rad) + y * np.sin(angle_rad)
        newy = y * np.cos(angle_rad) - x * np.sin(angle_rad)

        return newx, newy

    def rotate_bundle(self, angle: int = 90):

        # rotate the fiber data
        data = self.simbmap[['raoff', 'decoff']].copy()
        xx, yy = self._rotate_data(data.to_numpy(), angle)
        self.simbmap.raoff = xx  # r
        self.simbmap.decoff = yy  # d

        # rotate the hex offsets and recreate the sky hexagon
        hexx, hexy = self._rotate_data(self._hexagon_offsets, angle)
        self._hexagon_offsets = np.array(list(zip(hexx, hexy)))
        self.hexagon = self.coord.spherical_offsets_by(hexx * u.arcsec, hexy * u.arcsec)
        self.sky_region = PolygonSkyRegion(vertices=self.hexagon)

        self.angle += angle

    def plot(self, hdu: fits.ImageHDU = None, wcs: WCS = None, nx: int = 100,
             scale: float = None, sunit: str = 'arcsec', degaxis: bool = True,
             overlay_fibers: bool = False, **kwargs):

        if hdu and not wcs:
            wcs = WCS(hdu.header)
        elif wcs and not hdu:
            ny = nx
            hdu = fits.ImageHDU(np.ones([nx, ny]))
        elif not hdu:
            scale = scale or self.pixel_scale
            ny = nx
            hdu = fits.ImageHDU(np.ones([nx, ny]))
            wcs_dict = {
                'CTYPE1': 'RA---TAN', 'CUNIT1': 'deg', 'CD1_1': -2.77778e-4 * scale,
                'CRPIX1': nx // 2 + 1, 'CRVAL1': self.ra, 'NAXIS1': nx,
                'CTYPE2': 'DEC--TAN', 'CUNIT2': 'deg', 'CD2_2': 2.77778e-4 * scale,
                'CRPIX2': ny // 2 + 1, 'CRVAL2': self.dec, 'NAXIS2': ny}
            wcs = WCS(wcs_dict)

        plt.clf()
        if wcs.has_celestial:
            ww = wcs.celestial
        ax = plt.subplot(projection=ww)

        if degaxis:
            ax.coords[0].set_format_unit(u.degree, decimal=True)
            ax.coords[1].set_format_unit(u.degree, decimal=True)

        ax.imshow(hdu.data, **kwargs)

        # plot hexagon
        pp = self.sky_region.to_pixel(ww)
        pp.plot(ax=ax, edgecolor=kwargs.get('edgecolor', 'red'))

        # plot fibers
        if overlay_fibers:
            for fiber in self.simbmap[['raoff', 'decoff']][:self.size].iterrows():
                fiber_coord = self.coord.spherical_offsets_by(fiber[1]['raoff'] * u.arcsec,
                                                              fiber[1]['decoff'] * u.arcsec)
                p = SphericalCircle((fiber_coord.ra, fiber_coord.dec),
                                    scale / 2 * getattr(u, sunit),
                                    edgecolor='green', facecolor='none',
                                    transform=ax.get_transform('fk5'))
                ax.add_patch(p)

        return ax

    def create_ds9_region(self):
        return self.sky_region.serialize('ds9')

    def print_bundle(self):
        ''' Print the bundle to an Astropy Table '''

        tt = self.fibers.to_table()
        tt.add_column(np.arange(self.size) + 1, index=0, name='fiberid')
        ascii.write(tt, format='fixed_width_two_line',
                    formats={'ra': '{0:.12f}', 'dec': '{0:.12f}'})

    def print_hexagon(self):
        ''' Print the hexagon to an Astropy Table '''

        tt = self.hexagon.to_table()
        ascii.write(tt, format='fixed_width_two_line',
                    formats={'ra': '{0:.12f}', 'dec': '{0:.12f}'})


class MangaBundle(BaseBundle):
    survey = 'manga'
    pixel_scale = 2.5
    angle = 0.0
    nfiber = 127
    simfile = 'manga_simbmap_127.dat'


class LVMBundle(BaseBundle):
    survey = 'lvm'
    pixel_scale = 35.3
    angle = -90.0
    nfiber = 1801
    simfile = 'lvm_simbmap_1801.dat'


def plot_bundles(feet: list, ax=None, ra=None, dec=None, nx=100):
    """ plot a series of hex footprints or regions on a blank wcs """

    n = len(feet)

    # used only to generate a blank wcs
    if not ax:
        b = LVMBundle(ra=ra, dec=dec, ring=7)
        ax = b.plot(nx=nx)

    color = iter(plt.cm.rainbow(np.linspace(0, 1, n)))
    for foot in feet:
        reg = PolygonSkyRegion(vertices=SkyCoord(foot, unit=u.degree))
        pp = reg.to_pixel(ax.wcs)
        pp.plot(ax=ax, edgecolor=next(color))
